src/AGCertificateVerifier.cpp

```diff
@@ -0,0 +1,973 @@
+                    // Value is in next token
+                    jsmntok_t *value_tok = key_tok + 1;
+                    if (value_tok->type == JSMN_ARRAY) {
+                        // Next `value_tok->size' tokens are array elements
+                        jsmntok_t *t = value_tok + 1;
+                        jsmntok_t *endpos_tok = t + value_tok->size;
+                        while (t != endpos_tok) {
+                            if (t->type == JSMN_STRING) {
+                                const std::string spki = std::string(json + t->start, (size_t) t->end - t->start);
+                                revokedSHADigests.insert(spki);
+                            } else {
+                                break;
+                            }
+                            ++t;
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+    }
+    pos += json_len;
+
+    // Read CRLSet
+    unsigned char *end = (unsigned char *) (crlSetContent.c_str() + len);
+#define BCHECK(x) do { if (!(x)) break; } while (0)
+    while (pos < end) {
+        // Issuer SPKI SHA256 - 32 bytes
+        BCHECK(pos + 32 > end);
+        std::string issuerSHADigest = AGStringUtils::encodeToBase64(pos, 32);
+        pos += 32;
+        BCHECK(pos >= end);
+        // Serial number count - 4 bytes
+        uint32_t serialCount = *(uint32_t *)(pos);
+        pos += 4;
+        BCHECK(pos >= end);
+        // Serial number list
+        std::set<std::string> &serialList = issuerSHADigestToCRL[issuerSHADigest];
+        for (int i = 0; i < serialCount; i++) {
+            // Serial number size - 1 byte
+            uint8_t serialSize = *pos++;
+            BCHECK(pos >= end);
+            BCHECK(pos + serialSize > end);
+            // Serial number
+            serialList.insert(AGStringUtils::encodeToHex(pos, serialSize));
+            pos += serialSize;
+            BCHECK(pos >= end);
+        }
+    }
+#undef BCHECK
+}
+
+/**
+ * Update HPKP pinning information
+ * @param dnsName Hostname
+ * @param certChain Certificate chain
+ * @param header HTTP header name
+ * @param value HTTP header value
+ */
+void AGCertificateVerifier::updateHPKPInfo(
+        const std::string &dnsName, STACK_OF(X509) *certChain, const std::string &header, const std::string &value)
+{
+    if (AGStringUtils::toLower(value).find("report-only") != std::string::npos) {
+        // TODO: report
+        return;
+    }
+    AGHPKPInfo info = AGHPKPInfo(value);
+    if (!info.isValid()) {
+        return;
+    }
+    // HPKP info must have at least one pin in current chain and at least one pin not in current chain
+    if (!info.hasPinsNotInChain(certChain)) {
+        return;
+    }
+    if (!info.hasPinsInChain(certChain)) {
+        return;
+    }
+    dynamicHPKPInfo[dnsName] = info;
+    saveDynamicHPKPInfo();
+}
+
+/**
+ * Verify that certificate chain has no untrusted authorities
```

Comment:
It's not clear what exactly does this method do. What kind of untrusted authorities? Why can't they be in the chain?